src/utils/dockerDirManager.ts

```diff
@@ -0,0 +1,302 @@
+    }
+  }
+
+  /**
+   * Writes the docker CLI config.json file.
+   * @param config An object that is the config we want to write.
+   */
+  protected async writeDockerConfig(config: PartialDockerConfig): Promise<void> {
+    const rawConfig = JSON.stringify(config);
+
+    await fs.promises.mkdir(this.dockerDirPath, { recursive: true });
+    await fs.promises.writeFile(this.dockerConfigPath, rawConfig, { encoding: 'utf-8' });
+  }
+
+  /**
+   * Read the docker configuration, and return the docker socket in use by the
+   * current context.  If the context is invalid, return the default socket
+   * location.
+   * @param currentContext Docker's current context, as set in the configs.
+   */
+  protected async getCurrentDockerSocket(currentContext?: string): Promise<string> {
+    const defaultSocket = `unix://${ this.defaultDockerSockPath }`;
+
+    if (!currentContext) {
+      return defaultSocket;
+    }
+
+    for (const dir of await fs.promises.readdir(this.dockerContextDirPath)) {
+      const contextPath = path.join(this.dockerContextDirPath, dir, 'meta.json');
+
+      try {
+        const data = yaml.parse(await fs.promises.readFile(contextPath, 'utf-8'));
+
+        if (data.Name === currentContext) {
+          return data.Endpoints?.docker?.Host as string ?? defaultSocket;
+        }
+      } catch (ex) {
+        console.log(`Failed to read context ${ dir }, skipping: ${ ex }`);
+      }
+    }
+
+    // If we reach here, the current context is invalid.
+    return defaultSocket;
+  }
+
+  /**
+   * Given some information about state external to this method, returns the
+   * name of the context that should be used. Follows these rules, in order of preference:
+   * 1. If we have control of the default socket (`/var/run/docker.sock`), we should set the
+   *    context to it (which is actually un-setting the `currentContext` key).
+   *    This should have the widest compatibility.
+   * 2. Otherwise, check the current context and don't change anything if any of the following
+   *    is true:
+   *    - The current context uses a valid unix socket - the user is probably using it.
+   *    - The current context uses a non-unix socket (e.g. tcp) - we can't check if it's valid.
+   * 3. The current context is invalid - set the current context to our (rancher-desktop) context.
+   * @param weOwnDefaultSocket Whether Rancher Desktop has control over the default socket.
+   * @param currentContext The current context.
+   */
+  async getDesiredDockerContext(weOwnDefaultSocket: boolean, currentContext: string | undefined): Promise<string | undefined> {
+    if (weOwnDefaultSocket) {
+      return undefined;
+    }
+
+    if (!currentContext && !weOwnDefaultSocket) {
+      return this.contextName;
+    }
+
+    if (currentContext === this.contextName) {
+      return this.contextName;
+    }
+
+    const currentSocketUri = await this.getCurrentDockerSocket(currentContext);
+
+    if (!currentSocketUri.startsWith('unix://')) {
```

Comment:
Does this make sense on Windows? I believe we could still end up here, even though it's named pipes there (so as long as nobody is actively using it we can grab it, as if it's a tmpfs)…

Reply:
Looking at this again, I think that a better place to call `DockerDirManager.ensureDockerConfig` would be inside `WSLBackend`. Then, we can call it with only `true` as an argument, which will mean that we never get past the first `if` statement in `getDesiredDockerContext`.

Of course, that's already the case: `suppressSudo` will always be `false` on Windows (unless I'm missing something?), so we never get past the first `if` statement already. It's just better to call this in `WSLBackend` I think.

I'm going to make changes so that it's called in `WSLBackend`.

Reply:
The code is always correct, but it's from non-obvious code flow.   It would be good to either have this documented, or have an assertion (i.e. something like `if (os.platform() === 'win32') { throw new Error('blah'); }`).  Preferring the assertion because that means we change the caller forgetting about this we can see the failure.